src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -319,49 +322,92 @@ public Task<RewindData> GetRewindData(int height)
+
+                    if (targetHeight <= 0)
                     {
+                        transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
                         transaction.RemoveAllKeys("Coins", true);
+                        transaction.RemoveAllKeys("Rewind", true);
                         this.SetBlockHash(transaction, this.network.GenesisHash);
 
-                        res = this.network.GenesisHash;
+                        return this.network.GenesisHash;
                     }
-                    else
+                }
+
+                RewindData firstRewindData = null;
+
+                while (targetHeight < currentHeight)
+                {
+                    // Do a batch of blocks.
+                    var keysToRemove = new List<int>();
+                    var transactionsToRemove = new List<uint256>();
+                    var outputsToRestore = new List<UnspentOutputs>();
+
+                    using (DBreeze.Transactions.Transaction transaction = this.CreateTransaction())
                     {
                         transaction.ValuesLazyLoadingIsOn = false;
+                        int firstBlockToRewind = Math.Max(currentHeight - MaxBlocksToRewind + 1, (int)targetHeight + 1);
 
-                        Row<int, byte[]> firstRow = transaction.SelectBackward<int, byte[]>("Rewind").FirstOrDefault();
-                        transaction.RemoveKey("Rewind", firstRow.Key);
-                        var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(firstRow.Value);
-                        this.SetBlockHash(transaction, rewindData.PreviousBlockHash);
-
-                        foreach (uint256 txId in rewindData.TransactionsToRemove)
+                        for (int height = firstBlockToRewind; height <= currentHeight; height++)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
-                            transaction.RemoveKey("Coins", txId.ToBytes(false));
+                            Row<int, byte[]> row = transaction.Select<int, byte[]>("Rewind", height);
+                            if (!row.Exists)
+                                continue;
+
+                            keysToRemove.Add(row.Key);
+                            var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(row.Value);
+
+                            if (height == (targetHeight + 1))
+                                firstRewindData = rewindData;
+
+                            foreach (uint256 txId in rewindData.TransactionsToRemove)
+                            {
+                                this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
+                                transactionsToRemove.Add(txId);
+                            }
+
+                            foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
```

Comment:
I am not sure this is correct, you'll be adding to `outputsToRestore` the same trx with a different view of the chain.

We should probably call the actual dbreeze methods  `transaction.Insert("Coins"` and `transaction.RemoveKey("Coins",` on each iteration of this loop. Just mashing all the `Coins` in one list maybe problematic.

Reply:
Naive memory-based buffering should give the same results unless you see a circumstance where the in-memory data should also be updated. Will give this some more thought.